src/video_core/renderer_vulkan/vk_pipeline_serialization.cpp

```diff
@@ -0,0 +1,565 @@
+// SPDX-FileCopyrightText: Copyright 2025 shadPS4 Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "common/config.h"
+#include "common/hash.h"
+#include "shader_recompiler/frontend/fetch_shader.h"
+#include "shader_recompiler/info.h"
+#include "video_core/renderer_vulkan/vk_instance.h"
+#include "video_core/renderer_vulkan/vk_pipeline_cache.h"
+#include "video_core/renderer_vulkan/vk_pipeline_storage.h"
+#include "video_core/renderer_vulkan/vk_shader_util.h"
+
+namespace Serialization {
+
+/* You should increment versions below once corresponding serialization scheme is changed. */
+static constexpr u32 ShaderBinaryVersion = 0u;
+static constexpr u32 ShaderMetaVersion = 0u;
+static constexpr u32 PipelineKeyVersion = 0u;
+
+struct Archive {
+    void Alloc(size_t size) {
+        container.resize(size);
+    }
+
+    void Grow(size_t size) {
+        container.resize(container.size() + size);
+    }
+
+    void Merge(const Archive& ar) {
+        container.insert(container.end(), ar.container.cbegin(), ar.container.cend());
+        offset = container.size();
+    }
+
+    size_t SizeBytes() const {
+        return container.size();
+    }
+
+    u8* CurrPtr() {
+        return container.data() + offset;
+    }
+
+    void Advance(size_t size) {
+        assert((offset + size) < container.size());
+        offset += size;
+    }
+
+    std::vector<u8>&& TakeOff() {
+        offset = 0;
+        return std::move(container);
+    }
+
+    bool IsEoS() const {
+        return offset >= container.size();
+    }
+
+    Archive() = default;
+    explicit Archive(std::vector<u8>&& v) : container{v} {}
+
+    u32 offset{};
+    std::vector<u8> container{};
+};
+
+struct Writer {
+    template <typename T>
+    void Write(const T* ptr, size_t size) {
+        if ((ar.offset + size) >= ar.container.size()) {
+            ar.Grow(size);
```

Comment:
```suggestion
            ar.Grow(size + 1);
```
I'm not sure if this is the correct solution, but it does stop the assert in the Advance call 3 lines below from triggering on Debug builds.